op-acceptance-tests/tests/jovian/da_footprint_test.go

```diff
@@ -34,62 +40,192 @@ func (s *CalldataSpammer) Spam(t devtest.T) error {
 	return err
 }
 
+type daFootprintSystemConfig struct {
+	SetDAFootprintGasScalar func(scalar uint16) bindings.TypedCall[any] `sol:"setDAFootprintGasScalar"`
+	DAFootprintGasScalar    func() bindings.TypedCall[uint16]           `sol:"daFootprintGasScalar"`
+}
+
+type daFootprintL1Block struct {
+	DAFootprintGasScalar func() bindings.TypedCall[uint16] `sol:"daFootprintGasScalar"`
+}
+
+type daFootprintEnv struct {
+	l1Client     *dsl.L1ELNode
+	l2Network    *dsl.L2Network
+	l2EL         *dsl.L2ELNode
+	systemConfig daFootprintSystemConfig
+	l1Block      daFootprintL1Block
+}
+
+func newDAFootprintEnv(t devtest.T, l2Network *dsl.L2Network, l1EL *dsl.L1ELNode, l2EL *dsl.L2ELNode) *daFootprintEnv {
+	systemConfig := bindings.NewBindings[daFootprintSystemConfig](
+		bindings.WithClient(l1EL.EthClient()),
+		bindings.WithTo(l2Network.Escape().Deployment().SystemConfigProxyAddr()),
+		bindings.WithTest(t),
+	)
+
+	l1Block := bindings.NewBindings[daFootprintL1Block](
+		bindings.WithClient(l2EL.Escape().EthClient()),
+		bindings.WithTo(common.HexToAddress("0x4200000000000000000000000000000000000015")),
+		bindings.WithTest(t),
+	)
+
+	return &daFootprintEnv{
+		l1Client:     l1EL,
+		l2Network:    l2Network,
+		l2EL:         l2EL,
+		systemConfig: systemConfig,
+		l1Block:      l1Block,
+	}
+}
+
+func (env *daFootprintEnv) checkCompatibility(t devtest.T) {
+	// Ensure getters exist on both L1 SystemConfig and L2 L1Block
+	_, err := contractio.Read(env.systemConfig.DAFootprintGasScalar(), t.Ctx())
+	t.Require().NoError(err)
+	_, err = contractio.Read(env.l1Block.DAFootprintGasScalar(), t.Ctx())
+	t.Require().NoError(err)
+}
+
+func (env *daFootprintEnv) getSystemConfigOwner(t devtest.T) *dsl.EOA {
+	priv := env.l2Network.Escape().Keys().Secret(devkeys.SystemConfigOwner.Key(env.l2Network.ChainID().ToBig()))
+	return dsl.NewKey(t, priv).User(env.l1Client)
+}
+
+func (env *daFootprintEnv) setDAFootprintScalarViaSystemConfigOnL1(t devtest.T, scalar uint16) {
+	owner := env.getSystemConfigOwner(t)
+	_, err := contractio.Write(env.systemConfig.SetDAFootprintGasScalar(scalar), t.Ctx(), owner.Plan())
+	t.Require().NoError(err, "SetDAFootprintGasScalar transaction failed")
+	t.Logf("Set DA footprint gas scalar on L1: scalar=%d", scalar)
+}
+
+// waitForDAFootprintConfigChangeOnL2 waits until the L2 L1Block reports the expected scalar.
+func (env *daFootprintEnv) waitForDAFootprintConfigChangeOnL2(t devtest.T, expected uint16) {
+	t.Require().Eventually(func() bool {
+		current, err := contractio.Read(env.l1Block.DAFootprintGasScalar(), t.Ctx())
+		if err != nil {
+			return false
+		}
+		return current == expected
+	}, time.Minute, 2*time.Second, "L2 DA footprint gas scalar did not sync within timeout")
```

Comment:
I'm always a bit suspicious of arbitrary timeouts like `time.Minute`. These timeouts are especially flaky on persistent devnets where everything is slower ([here](https://github.com/ethereum-optimism/optimism/pull/17772) is an example).

Rather than waiting an arbitrary length of time for the value to be updated, what do you think about asserting that it's updated in the first L2 block of the correct epoch? I think that would make the test more robust and remove timing assumptions.

Reply:
The minute wasn't really arbitrary but chosen to give the L1 block enough time to become the L1 origin.

But I improved the test, to first set the scalar, then wait until the set-tx's receipt block number becomes the L1 origin, which is the first L2 block that has the new scalar.